File: archive/f/f-sharp/JosephusProblem.fs
open System

module Josephus =
    let run (n, k) =
        if n = 1 then
            Ok 1
        elif k = 1 then
            Ok n
        elif k = 2 then
            // fast formula for k = 2
            let mutable p = 1

            while p <= n do
                p <- p <<< 1

            Ok(2 * (n - (p >>> 1)) + 1)
        else
            // general k > 2 using iterative approach
            let mutable survivor = 0

            for i in 2..n do
                survivor <- (survivor + k) % i

            Ok(survivor + 1)

module Helpers =
    let usage =
        "Usage: please input the total number of people and number of people to skip."

    let private (|Int|_|) (s: string) =
        match s.Trim() |> Int32.TryParse with
        | true, n -> Some n
        | false, _ -> None

    let parseArgs argv =
        match argv with
        | [| Int n; Int k |] when n > 0 && k > 0 -> Ok(n, k)
        | _ -> Error usage

    let handleResult =
        function
        | Ok result ->
            printfn "%d" result
            0
        | Error msg ->
            eprintfn "%s" msg
            1

[<EntryPoint>]
let main argv =
    argv |> Helpers.parseArgs |> Result.bind Josephus.run |> Helpers.handleResult

File: archive/t/typescript/josephus-problem.ts
function printUsage(): void {
  console.error(
    "Usage: please input the total number of people and number of people to skip.",
  );
  process.exit(1);
}

function josephus(n: number, k: number): number {
  if (n === 1) return 1;
  return ((josephus(n - 1, k) + k - 1) % n) + 1;
}

function main(): void {
  const n = Number.parseInt(process.argv[2], 10);
  const k = Number.parseInt(process.argv[3], 10);

  if (Number.isNaN(n) || Number.isNaN(k) || n <= 0 || k <= 0) {
    printUsage();
  }

  console.log(josephus(n, k).toString());
}

main();
